crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -120,6 +136,83 @@ where
             is_valid
         })
     }
+
+    /// Endless future that updates the reorged logs for impacted filters
+    async fn update_reorged_logs<Events, Network>(
+        &self,
+        pubsub: EthPubSub<Provider, Pool, Events, Network>,
+    ) where
+        Events: CanonStateSubscriptions + 'static,
+        Network: NetworkInfo + 'static,
+    {
+        let mut stream = self.reorged_logs_stream(pubsub).await;
+        while let Some((id, mut logs)) = stream.next().await {
+            let mut filters = executor::block_on(self.active_filters().inner.lock());
+            let active_filter =
+                filters.get_mut(&id).ok_or(FilterError::FilterNotFound(id)).unwrap();
+            let mut guard = active_filter.reorged_logs.lock().await;
+            if let Some(reorged_logs_vec) = guard.as_mut() {
+                reorged_logs_vec.append(&mut logs);
+            }
+        }
+    }
+
+    /// Reacts to reorged blocks, checks impacted log filters, stores reorged logs in the log filter
+    pub async fn reorged_logs_stream<Events, Network>(
+        &self,
+        pubsub: EthPubSub<Provider, Pool, Events, Network>,
+    ) -> impl Stream<Item = (FilterId, Vec<Log>)> + '_
+    where
+        Events: CanonStateSubscriptions + 'static,
+        Network: NetworkInfo + 'static,
+    {
+        let mut temp_reorged_blocks = Vec::new();
+
+        BroadcastStream::new(pubsub.get_chain_events().subscribe_to_canonical_state())
+            .map(move |canon_state| {
+                canon_state.expect("new block subscription never ends; qed").block_receipts()
+            })
+            .flat_map(futures::stream::iter)
+            .flat_map(move |(block_receipts, removed)| {
+                let mut reorged_logs: Vec<(FilterId, Vec<Log>)> = Vec::new();
+
+                if removed {
+                    temp_reorged_blocks.push(block_receipts);
+
+                    let filters = executor::block_on(self.active_filters().inner.lock());
```

Comment:
no executor block,

this needs to be refactored so we can have async code

I suggest not do the flat_map stuff and instead move all of this to async code in the while loop